### PR TITLE
Be lazy in choosing which handler to use

### PR DESCRIPTION
**Pull request by michaelfig**
This fixes a race between forwarding a promise and discarding its old handler.

### CODE REVIEW

**Comment by michaelfig** [comment]
@warner, here is the test failure, for your convenience:

```
    expected:
      - pipe1
      - pipe2
      - pipe3
      - p1.then
      - p2.then
      - p3.then
    actual:
      - pipe1
      - p1.then
      - pipe2
      - p2.then
      - pipe3
      - p3.then
    at: Test.<anonymous>
      (/home/runner/work/agoric-sdk/agoric-sdk/packages/SwingSet/test/test-message-patterns.js:67:9)
    stack: >-
      Error: should be equivalent
          at Test.assert [as _assert] (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/tape/lib/test.js:77:32)
          at Test.tapeDeepEqual (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/tape/lib/test.js:422:10)
          at Test.bound [as deepEqual] (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/home/runner/work/agoric-sdk/agoric-sdk/packages/SwingSet/test/test-message-patterns.js:67:9)
 ```

**Comment by warner** [comment]
Let's hold off landing this for now, I'm not sure it's doing what I want. (this is not your fault at all.. I'm still investigating how ordering works, and what liveslots ought to be doing, so I'm not sure that I'm asking for the right thing yet, and the previous code might work better for me than this change, once I fix liveslots to do the right thing)

**Review by erights — APPROVED** [review]
LGTM! (Once you confirm that my explanation in keybase chat is accurate)

More explanatory text somewhere would be useful, so it is easier to figure such things out. But they can be in a separate doc rather than in this PR.

**Review by warner — APPROVED** [review]
These changes looks good to me. I've rebased and incorporated this commit into my #1068 PR, which also updates the swingset tests (which were failing in this PR because they need to match the new HP behavior). So don't land this one, but rather I'll close this one when 1068 lands.

**Pull request closed without merging**
